arch-riscv: Relation chain on RVV support

aarmejach opened the pull request:
Gerrit relation chain on RVV support found here:
https://gem5-review.googlesource.com/c/public/gem5/+/68418/15

huxuan0307 committed:
arch-riscv: Add risc-v vector ext v1.0 vset insts support

Change-Id: I84363164ca327151101e8a1c3d8441a66338c909
Co-authored-by: Yang Liu <numbksco@gmail.com>
Co-authored-by: Fan Yang <1209202421@qq.com>

arch-riscv: Add a todo to fix vsetvl stall on decode

Change-Id: Iafb129648fba89009345f0c0ad3710f773379bf6

huxuan0307 committed:
arch-riscv: Add risc-v vector ext v1.0 mem insts support

* TODOs:
  + Vector Segment Load/Store
  + Vector Fault-only-first Load

Change-Id: I2815c76404e62babab7e9466e4ea33ea87e66e75
Co-authored-by: Yang Liu <numbksco@gmail.com>
Co-authored-by: Fan Yang <1209202421@qq.com>
Co-authored-by: Jerin Joy <joy@rivosinc.com>

huxuan0307 committed:
arch-riscv: Add risc-v vector ext v1.0 arith insts support

TODOs:
  + vcompress.vm

Change-Id: I86eceae66e90380416fd3be2c10ad616512b5eba
Co-authored-by: Yang Liu <numbksco@gmail.com>
Co-authored-by: Fan Yang <1209202421@qq.com>
Co-authored-by: Jerin Joy <joy@rivosinc.com>

arch-riscv: Add LICENCE to template files

Change-Id: I825e72bffb84cce559d2e4c1fc2246c3b05a1243

huxuan0307 committed:
arch-riscv: Add risc-v vector regs and configs

This commit add regs and configs for vector extension

* Add 32 vector arch regs as spec defined and 8 internal regs for
  uop-based vector implementation.
* Add default vector configs(VLEN = 256, ELEN = 64). These cannot
  be changed yet, since the vector implementation has only be tested
  with such configs.
* Add disassamble register name v0~v31 and vtmp0~vtmp7.
* Add CSR registers defined in RISCV Vector Spec v1.0.
* Add vector bitfields.
* Add vector operand_types and operands.

Change-Id: I7bbab1ee9e0aa804d6f15ef7b77fac22d4f7212a
Co-authored-by: Yang Liu <numbksco@gmail.com>
Co-authored-by: Fan Yang <1209202421@qq.com>
Co-authored-by: Jerin Joy <joy@rivosinc.com>

arch-riscv: enable rvv flags only for RV64

Change-Id: I6586e322dfd562b598f63a18964d17326c14d4cf

aarmejach commented:
The test that is failing is: `test-gem5-library-riscv-hello-restore-checkpoint-ALL-x86_64-opt`

This test is meant to fail given the changes on the vector register file done on these commits. Should we disable it?

powerjg commented:
We should probably fix it before we merge this change.

I think that we'll need to create a new checkpoint and update the test to use that new checkpoint. Give us a couple of days to look into this. @BobbyRBruce and @mkjost0

BobbyRBruce commented:
I'll work on a solution to this this afternoon. It shouldn't take too long.

These checkpoint tests are quite an overhead to maintain ...

powerjg commented:
@BobbyRBruce, don't forget to work on this checkpoint "fix"

BobbyRBruce commented:
All you need to do to get the tests passing is change this line https://github.com/gem5/gem5/blob/3c6563d6f7cf35f5e801b7a1dc9363619bb55239/configs/example/gem5_library/checkpoints/riscv-hello-restore-checkpoint.py#L93 to `checkpoint=Resource("riscv-hello-example-checkpoint")`.

Though, in doing this, I discovered the "util/cpt_upgrader.py" did not work in this case. I logged this in Issue #106. I'm not familiar with the upgrader, but we'll probably need to fix this issue as we want to give people an easy way to upgrade their checkpoints to work with this.

powerjg commented:
I think it would be good to implement the checkpoint upgrader #106 before merging this change. If you have questions on how to do this, let me know and I can try to help (I'm also not an expert 😄 ).

adriaarmejach committed:
configs: update riscv restore checkpoint test

Change-Id: I019fc6394a03196711ab52533ad8062b22c89daf

aarmejach commented:
Thanks Bobby, I have made the modification to the config file to use the new checkpoint restore test. Let's see if tests pass now.

I can help on the cpt_upgrader if you want, I have an idea on what needs to be done there and can take a look at it on Monday.

P.S: I had way too many develop branches on diferent remotes and made a bit of a mess on my first push, think it is fixed now.

rogerchang23424 commented on src/arch/riscv/isa.cc:
```diff
@@ -299,6 +314,8 @@ void ISA::clear()
     // mark FS is initial
     status.fs = INITIAL;
 
+    status.vs = VPUStatus::INITIAL;
```
```suggestion
    if (rv_type == RV64) {
        status.vs = VPUStatus::INITIAL;
        misa.rvv = 1;
    }
```

rogerchang23424 replied:
I suggest it because the vector only implemented for RV64 now

aarmejach replied:
done, added to the switch statement.

vspetrov commented:
Hello, we've been looking into this patch and faced an issue with O3 CPU. The reproducer is quite large and depends on our internal library - we are working on making it smaller and plan to share it here. Meanwhile, i wanted to share our current view and check if the understanding of the problem is correct. 

Basically, the reproducer looks like this:
```
vsetvl (type1, vl1) //First vset vl instruction that sets custom type/vl for decode
inst_squash // any instruction that will cause a pipeline squash
vse // Any vector instruction that relies on type1,vl1 durin decode
...
vsetvl (type2, vl2) // Second vset vl with different type/vl
vse // any vector instr that uses type2/vl2
```
What happens is: O3 CPU starts executing vsetvl №1 and saves type1/vl1 on decoder structure. Then It would create inst_squash - this can be any instruction that will cause a pipeline squash in future (it can be ld/st speculation or branch for example). Then vsetvl 2 is executed BEFORE inst_squash actually triggers pipeline squash. THis saves new type/vl on decoder. This is when pipeline is actually squashed and restarted from inst_squash. However, decoding of 1st vse instruction now uses wrong type/vl values cached on decoder from 2nd vsetvl execution.

Does that view of the problem makes sense or we missing smth?

It looks like for such scenarios we might need to maintain kind of stack data structure that would be rolled back in case of pipeline squashes (smth similar to bpu history state stack). However, it will require some mechanism to pass that info from CPU to decoder. Maybe seqNum.

aarmejach commented:
The problem you describe makes sense to me. And the solution of rolling back decoder state on pipeline squashes is pobably a good way to fix it, and is related to ongoing discussion on how to better implement vsetvl instructions.

AFAIK the vsetvl instructions now stall younger instructions via a fake dependency, and there has been discussion to change vsetvl to behave like a control instruction (basically branch-like) to enable speculation around them. That is, speculate that the vl/vtype is the same as the previous and, if not, a rollback is done. Therefore, the described problem and the proposed solution can probably be used to rethink the vsetvl implementation and solve two problems in one shot.

To be honest i'd like to see this current PR merged ASAP, and try to make new issues describing problems we are seeing to fix them (like this vsetvl issue). This will allow much faster progress as now this PR is blocking multiple improvements. 

I'd like some thoughts on this. Of course we should go through code review, etc., but if we merge this PR it will enable faster progress.

BTW CI tests should pass now without problems.

powerjg commented:
My bad, @aarmejach . I'll make sure this is reviewed ASAP with a goal of getting it merged this week.

vspetrov commented:
@aarmejach thanks. sounds good. Just out of curiosity, do you know of any actual public RTL implementation of this vector extension and how the vsetvl states are handled there?

powerjg commented:
@vspetrov IIRC, the RVV standard states that `vsetvl` should be treated as a control instruction. Although, I don't have the exact citation for you.

From my understanding, if you're going to execute/schedule the vector instructions on an out-of-order core, the only option is to treat `vsetvl` as a control instruction. The performance will significantly suffer or correctness will be impacted, if you stall younger instructions after the `vsetvl`, at least with how many compilers generate code.

However, if you have a fully decoupled vector unit, then you would probably have more freedom in how you implement the `vsetvl` instruction (e.g., you could execute that instruction in the front-end). 

That said, the current implementation in gem5 is going to be using the core (out-of-order or otherwise) to execute the vector instructions. So, at least in the longer term, we will likely implement `vsetvl` as a control instruction.

In the short term, we should probably just fail (e.g., `fatal`) if trying to use the O3CPU with vector extensions enabled until we implement `vsetvl` in a more accurate way.

hnpl commented:
In case you're interested, the chapter 9 of the rvv spec has a note about control dependencies related to the vl register,

https://github.com/riscv/riscv-v-spec/blob/v1.0/v-spec.adoc#9-vector-memory-consistency-model

powerjg commented on src/arch/riscv/isa.cc:
```diff
@@ -234,11 +244,9 @@ namespace
+RegClass vecPredRegClass(VecPredRegClass, VecPredRegClassName, 0,
         debug::IntRegs);
-RegClass matRegClass(MatRegClass, MatRegClassName, 1, debug::MatRegs);
+RegClass matRegClass(MatRegClass, MatRegClassName, 0, debug::MatRegs);
```
Was this change purposeful or did it sneak in because of a rebase (the change is a 1 -> 0).

Same for `vecPredRegClass` above.

aarmejach replied:
There are no specific predicate/mask registers, so `vecPredRegClass` should be empty, i.e., `0` should be ok. 

Same for `matRegClass`, until a matrix extension shows up.

powerjg commented on src/arch/riscv/decoder.cc:
```diff
@@ -42,13 +42,21 @@ void Decoder::reset()
 void
 Decoder::moreBytes(const PCStateBase &pc, Addr fetchPC)
 {
+    if (GEM5_UNLIKELY(!this->vConfigDone)) {
```
Can we put a `TODO` here since we all agree that this is not a good long-term solution?

aarmejach replied:
done.

powerjg commented on src/arch/riscv/isa/templates/templates.isa:
This file (and the others in this directory) are missing a LICENSE. You should be able to copy the BSD license here.

aarmejach replied:
done

powerjg commented on src/cpu/minor/fetch2.cc:
I think I would actually prefer to drop the minor/o3 changes and instead say that they are unsupported in this change.

We could potentially put these to changes up on a separate PR and point people to that if they insist on using minor/o3 with RVV.

aarmejach replied:
I agree. Dropped the 3 commits that were just to make minor/o3 work using the current `vset*` implementation.

I can create a PR with these 3 commits once this PR is merged. Which will be just for people that wants to use minor/o3 while we fix the `vset*` implementation.

powerjg replied:
If you don't mind, I'm going to add a commit to this PR which throws a fatal if RVV is used with O3 or Minor CPU. (Assuming I can figure out how to do it!)

aarmejach replied:
Sure, please go ahead!

powerjg reviewed (commented):
Here's an initial review. I have a few more files I'd like to go through, but I don't expect any big hurdles.

Overall, this looks great. Let's get it in and improve on it from there!

rogerchang23424 commented on src/arch/riscv/regs/vector.hh:
```diff
@@ -68,12 +69,18 @@ const std::vector<std::string> VecRegNames = {
+inline constexpr RegClass vecElemClass =
+    RegClass(VecElemClass, VecElemClassName, NumVecRegs * NumVecElemPerVecReg,
+            debug::VecRegs).
+        ops(vecRegElemClassOps);
```
I think we should drop the vecElemClass commit because it's just solve the hanging problem for `vsetvl`, `vsetvli` and `vsetivli` in minor CPU.

aarmejach replied:
done, i removed this one and the other 2 commits specific to minor/o3.

The drawback is that now the `restore-checkpoint` test will fail again. This has 2 implications:

1.  @BobbyRBruce could you fix the test again, basically the `regs.vector_element` field in the cpt now should be empty.
2. I will amend the PR #115 on the cpt_upgrader now to take this change into account.

BobbyRBruce replied:
> done, i removed this one and the other 2 commits specific to minor/o3.
> 
> The drawback is that now the `restore-checkpoint` test will fail again. This has 2 implications:
> 
> 1. @BobbyRBruce could you fix the test again, basically the `regs.vector_element` field in the cpt now should be empty.
> 2. I will amend the PR [util: fix cpt upgrader for rvv changes in PR #83 #115](https://github.com/gem5/gem5/pull/115) on the cpt_upgrader now to take this change into account.

I've updated the restore checkpoint. It should work with the PR as it currently stands.

vmskv commented:
> Hello, we've been looking into this patch and faced an issue with O3 CPU. The reproducer is quite large and depends on our internal library - we are working on making it smaller and plan to share it here. Meanwhile, i wanted to share our current view and check if the understanding of the problem is correct.
> 
> Basically, the reproducer looks like this:
> 
> ```
> vsetvl (type1, vl1) //First vset vl instruction that sets custom type/vl for decode
> inst_squash // any instruction that will cause a pipeline squash
> vse // Any vector instruction that relies on type1,vl1 durin decode
> ...
> vsetvl (type2, vl2) // Second vset vl with different type/vl
> vse // any vector instr that uses type2/vl2
> ```
> 
> What happens is: O3 CPU starts executing vsetvl №1 and saves type1/vl1 on decoder structure. Then It would create inst_squash - this can be any instruction that will cause a pipeline squash in future (it can be ld/st speculation or branch for example). Then vsetvl 2 is executed BEFORE inst_squash actually triggers pipeline squash. THis saves new type/vl on decoder. This is when pipeline is actually squashed and restarted from inst_squash. However, decoding of 1st vse instruction now uses wrong type/vl values cached on decoder from 2nd vsetvl execution.
> 
> Does that view of the problem makes sense or we missing smth?
> 
> It looks like for such scenarios we might need to maintain kind of stack data structure that would be rolled back in case of pipeline squashes (smth similar to bpu history state stack). However, it will require some mechanism to pass that info from CPU to decoder. Maybe seqNum.

The shareable reproducer has been prepared:  
https://github.com/vmskv/gem5/tree/vmskv/rvv_bad_o3_vsetvl_execuion/riscv_o3_bad_vsetvl_execution_reproducer  
In out-of-order(O3CPU) mode if vsetvl is in mispredicted branch it changes the 'vl' value in decoder and the vector instructions afterwards are decoded/executed bad after the mispredicted vsetvl squashed.

aarmejach commented:
Thanks @vmskv. I think we should open an issue with the description of the problem and the reproducible, which will be the motivation to modify the current `vset*` implementation.

rogerchang23424 commented:
Thanks for the update.

Cloud you rebase the branch to the top of develop tree? It has the conflict with develop branch.

Thanks

powerjg committed:
arch-riscv: Add fatal if RVV used with o3 or minor

Since the O3 and Minor CPU models do not support RVV right now as the
implementation stalls the decode until vsetvl instructions are exectued,
this change calls `fatal` if RVV is not explicitly enabled.

It is possible to override this if you explicitly enable RVV in the
config file.

Change-Id: Ia801911141bb2fb2bedcff3e139bf41ba8936085
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg committed:
arch-riscv: Improve style

Minor style fixes in vector code

Change-Id: If0de45a2dbfb5d5aaa65ed3b5d91d9bee9bcc960
Signed-off-by: Jason Lowe-Power <jason@lowepower.com>

powerjg commented:
I went ahead and added a couple of commits to this chain. If it's OK with you, I can simply push them here. Otherwise, you can pull them from my repo: https://github.com/powerjg/gem5/commits/pr/aarmejach/83

I fixed some style problems and I added a fatal if RVV is used with o3 or minor.

aarmejach commented:
@powerjg feel free to push them here.

@BobbyRBruce a quick way to fix the restore-checkpoint test again is to run the cpt_upgrader of PR #115 over the checkpoint used in `checkpoint=Resource("riscv-hello-example-checkpoint")` that this PR is using.

Hopefully, this should get us close to merge this. I am going into vacation mode, but will monitor this to make sure we get it in asap.

rogerchang23424 commented on src/arch/riscv/isa/decoder.isa:
```diff
@@ -2012,6 +2280,2093 @@ decode QUADRANT default Unknown::unknown() {
+                        uint64_t requested_vtype = zimm11;
+
+                        Rd_ud = 0;
+                    }}, VectorConfigOp, IsDirectControl, IsCondControl);
```
```suggestion
                    }}, VectorConfigOp, IsSerializeAfter, IsNonSpeculative);
```
We should treat vset\*vl\* like csrrw instruction

aarmejach replied:
Yes, in this current implementation we can treat them like `csrrw`. In future speculative implementations we can go back to the branch-like semantics.

rogerchang23424 commented on src/arch/riscv/isa/decoder.isa:
```diff
@@ -2012,6 +2280,2093 @@ decode QUADRANT default Unknown::unknown() {
+                            uint64_t requested_vtype = Rs2_ud;
+
+                            Rd_ud = 0;
+                        }}, VectorConfigOp, IsDirectControl, IsCondControl);
```
ditto

rogerchang23424 commented on src/arch/riscv/isa/decoder.isa:
```diff
@@ -2012,6 +2280,2093 @@ decode QUADRANT default Unknown::unknown() {
+                            uint64_t requested_vtype = zimm10;
+
+                            Rd_ud = 0;
+                        }}, VectorConfigOp, IsDirectControl, IsCondControl);
```
ditto

vmskv commented:
> Thanks @vmskv. I think we should open an issue with the description of the problem and the reproducible, which will be the motivation to modify the current `vset*` implementation.

The issue is opened https://github.com/gem5/gem5/issues/144

rogerchang23424 commented:
Hi @powerjg,

The additional commits are LGTM. However, I think it will be better to move the `enable_rvv` option from `RiscvDecorder` to `RiscvISA` as it is one of isa configuration. There are two benefits of this.

1. Some of isa options `riscv_type` and `check_alignment` are put in `RiscvISA`. If the `enable_rvv` is in `RiscvISA`, it will be easier for user to configure.
2. The initialization of `mstatus` and `misa` CSR are depending on the `enable_rvv` option.

(nit) Riscv decoder can get `enable_rvv` option by `(dynamic_cast<ISA*>(isa))->enableRvv();` .

powerjg commented:
Thanks for the feedback @rogerchang23424, I took your suggestions into account.

@aarmejach, I had to force push. I hope that doesn't cause a big headache for you.

There's one more comment to address from Roger and we need to figure out the checkpoint. After that, I think we're good to go.

adriaarmejach committed:
arch-riscv: Make vset*vl* instructions serialize

Current implementation of vset*vl* instructions serialize pipeline and
are non-speculative.

Change-Id: Ibf93b60133fb3340690b126db12827e36e2c202d

rogerchang23424 reviewed (approved):
Let Jason and Bobby do final review

aarmejach commented:
Sure, all comments are addressed and tests should pass.

Once a final review is done feel free to hit the button.

powerjg reviewed (approved):
Thanks @aarmejach for pushing this through!

powerjg commented:
Thanks for the hard work everyone!
